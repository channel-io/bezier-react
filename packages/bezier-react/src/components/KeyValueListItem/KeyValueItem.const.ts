export const TEST_ID_MAP = {
  ROOT: 'bezier-react-key-value-list-item',
  MULTILINE_ROOT: 'bezier-react-key-value-list-item-multiline',
  KEY_ITEM: 'bezier-react-key-value-list-item-key-item',
  VALUE_ITEM: 'bezier-react-key-value-list-item-value-item',
  ACTIONS_ITEM: 'bezier-react-key-value-list-item-actions-item',
}
