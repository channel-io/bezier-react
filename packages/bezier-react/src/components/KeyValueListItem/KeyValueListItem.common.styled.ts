/* Internal dependencies */
import React from 'react'
import { css, styled } from 'Foundation'
import { InterpolationProps } from 'Types/Foundation'

export const KeyValueListItemWrapper = styled.div<InterpolationProps>`
  ${({ foundation }) => foundation?.rounding?.round6}
  ${({ foundation }) => foundation?.transition.getTransitionsCSS(['background-color', 'color'])}
  ${({ interpolation }) => interpolation}
`

export interface KeyValueListItemContainerProps {
  onClick?: (e: React.MouseEvent<HTMLDivElement>) => void
}

export const KeyValueListItemContainer = styled.div<KeyValueListItemContainerProps>`
  box-sizing: border-box;
  display: flex;
  align-items: flex-start;
  padding: 4px 6px;

  ${({ foundation }) => foundation?.rounding?.round6};
  ${({ onClick }) => onClick && css`
    cursor: pointer;
    &:hover {
      background-color: ${({ foundation }) => foundation?.theme?.['bg-black-lighter']};
    }
  `};
`
