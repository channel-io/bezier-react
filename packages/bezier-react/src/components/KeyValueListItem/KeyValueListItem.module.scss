.KeyValueItemWrapper {
  border-radius: var(--radius-6);
  transition: background-color var(--transition-s), color var(--transition-s);
}

.KeyValueItemInnerWrapper {
  display: flex;
  align-items: flex-start;

  box-sizing: border-box;
  padding: 4px 6px;

  border-radius: var(--radius-6);

  &:where(.clickable) {
    cursor: pointer;

    &:hover {
      background-color: var(--bg-black-lighter);
    }
  }
}

.KeyValueSingleLineItem {
  display: flex;
  align-items: flex-start;
}

.KeySingleLineItem {
  overflow: hidden;
  flex: 1;
}

.ValueSingleLineItem {
  overflow: hidden;
  flex: 2;
  justify-content: space-between;
  min-height: 28px;
}

.KeyItem {
  overflow: hidden;
  display: flex;
  align-items: center;
}

.KeyText {
  margin-left: 8px;
}

.ValueItem {
  width: 100%;

  /* To avoid inheritance issues with the internal `Text` color */
  color: var(--txt-black-darkest);
}

.ItemAction {
  display: flex;
  align-items: center;
}
