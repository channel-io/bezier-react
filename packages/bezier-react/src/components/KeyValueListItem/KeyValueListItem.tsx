import React, {
  type Ref,
  forwardRef,
} from 'react'

import { TEST_ID_MAP } from './KeyValueListItem.const'
import { type KeyValueListItemProps } from './KeyValueListItem.types'
import {
  ItemAction,
  KeyItem,
  ValueItem,
} from './common'

import * as Styled from './KeyValueListItem.styled'

function KeyValueListItem(
  {
    className,
    interpolation,
    valueWrapperInterpolation,
    keyWrapperInterpolation,
    keyIcon,
    keyContent,
    actions,
    testId = TEST_ID_MAP.ROOT,
    children,
    onClickKey,
    onClickValue,
    ...props
  }: KeyValueListItemProps,
  forwardedRef: Ref<HTMLDivElement>,
) {
  return (
    <Styled.Wrapper
      data-testid={testId}
      {...props}
      ref={forwardedRef}
      interpolation={interpolation}
      className={className}
    >
      <Styled.KeyItemContainer onClick={onClickKey}>
        <KeyItem
          keyIcon={keyIcon}
          interpolation={keyWrapperInterpolation}
        >
          { keyContent }
        </KeyItem>
      </Styled.KeyItemContainer>
      <Styled.ValueItemContainer onClick={onClickValue}>
        <ValueItem interpolation={valueWrapperInterpolation}>
          { children }
        </ValueItem>
        <ItemAction actions={actions} />
      </Styled.ValueItemContainer>
    </Styled.Wrapper>
  )
}

export default forwardRef(KeyValueListItem)
