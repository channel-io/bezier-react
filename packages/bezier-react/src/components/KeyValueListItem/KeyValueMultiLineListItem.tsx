/* External dependencies */
import React, {
  forwardRef,
  type Ref,
} from 'react'

/* Internal dependencies */
import {
  ValueItem,
  ItemAction,
  KeyItem,
} from './common'
import { TEST_ID_MAP } from './KeyValueListItem.const'
import { type KeyValueListItemProps } from './KeyValueListItem.types'
import * as Styled from './KeyValueMultiLineListItem.styled'

function KeyValueMultiLineListItem(
  {
    className,
    interpolation,
    valueWrapperInterpolation,
    keyWrapperInterpolation,
    keyIcon,
    keyContent,
    actions,
    testId = TEST_ID_MAP.MULTILINE_ROOT,
    children,
    onClickKey,
    onClickValue,
    ...props
  }: KeyValueListItemProps,
  forwardedRef: Ref<HTMLDivElement>,
) {
  return (
    <Styled.Wrapper
      data-testid={testId}
      {...props}
      ref={forwardedRef}
      interpolation={interpolation}
      className={className}
    >
      <Styled.KeyItemContainer onClick={onClickKey}>
        <KeyItem
          keyIcon={keyIcon}
          interpolation={keyWrapperInterpolation}
        >
          { keyContent }
        </KeyItem>
        <ItemAction actions={actions} />
      </Styled.KeyItemContainer>
      <Styled.ValueItemContainer onClick={onClickValue}>
        <ValueItem interpolation={valueWrapperInterpolation} multiline>
          { children }
        </ValueItem>
      </Styled.ValueItemContainer>
    </Styled.Wrapper>
  )
}

export default forwardRef(KeyValueMultiLineListItem)
