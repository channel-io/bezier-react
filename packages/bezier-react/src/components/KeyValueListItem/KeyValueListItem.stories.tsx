/* eslint-disable no-console */
/* External dependencies */
import React from 'react'
import base from 'paths.macro'
import {
  type Meta,
  type Story,
} from '@storybook/react'

/* Internal dependencies */
import { getTitle } from '~/src/utils/storyUtils'
import {
  Button,
  ButtonColorVariant,
  ButtonStyleVariant,
} from '~/src/components/Button'
import {
  BadgeIcon,
  EditIcon,
} from '~/src/components/Icon'
import KeyValueListItem from './KeyValueListItem'
import KeyValueMultiLineListItem from './KeyValueMultiLineListItem'
import { type KeyValueListItemProps } from './KeyValueListItem.types'

export default {
  title: getTitle(base),
  component: KeyValueListItem,
} as Meta

interface KeyValueListItemStorybookProps extends KeyValueListItemProps {
  containerWidth: number
}

const DEFAULT_ARGS: KeyValueListItemStorybookProps = {
  keyIcon: BadgeIcon,
  keyContent: 'I\'m the Long Long Key',
  containerWidth: 300,
  // eslint-disable-next-line max-len
  children: 'Lorem ipsum dolor sit, amet consectetur adipisicing elit. Perspiciatis eveniet minus nobis nemo quisquam ipsum ut. Iste molestiae sint, laboriosam minima, quaerat velit blanditiis natus quos non vel fugiat perspiciatis.',
}

function onClickKey() {
  console.log('Key')
}

function onClickValue() {
  console.log('Value')
}

const SingleLineTemplate: Story<KeyValueListItemStorybookProps> = ({ containerWidth, ...otherProps }) => (
  <div
    style={{
      width: `${containerWidth}px`,
    }}
  >
    <KeyValueListItem {...otherProps} />
    <KeyValueListItem
      {...otherProps}
      actions={Array.from(Array(2)).map(() => ({ icon: EditIcon, onClick: console.log, tooltip: '수정하기' }))}
    />
    <KeyValueListItem {...otherProps} onClickKey={onClickKey} onClickValue={onClickValue} />
    <KeyValueListItem {...otherProps}>
      <Button
        leftContent="play"
        text="높이 테스트"
        colorVariant={ButtonColorVariant.Green}
        styleVariant={ButtonStyleVariant.Primary}
      />
    </KeyValueListItem>
  </div>
)

const MultiLineTemplate: Story<KeyValueListItemStorybookProps> = ({ containerWidth, ...otherProps }) => (
  <div
    style={{
      width: `${containerWidth}px`,
    }}
  >
    <KeyValueMultiLineListItem {...otherProps} />
    <KeyValueMultiLineListItem
      {...otherProps}
      actions={Array.from(Array(2)).map(() => ({ icon: EditIcon, onClick: console.log, tooltip: '수정하기' }))}
    />
    <KeyValueMultiLineListItem {...otherProps} onClickKey={onClickKey} onClickValue={onClickValue} />
  </div>
)

export const SingleLine = SingleLineTemplate.bind({})
SingleLine.args = {
  ...DEFAULT_ARGS,
  actions: { icon: EditIcon, onClick: console.log, tooltip: '수정하기' },
}

export const MultiLine = MultiLineTemplate.bind({})
MultiLine.args = {
  ...DEFAULT_ARGS,
}
