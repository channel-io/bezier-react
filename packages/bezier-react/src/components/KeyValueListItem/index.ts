import KeyValueListItem from './KeyValueListItem'
import type { KeyValueListItemProps } from './KeyValueListItem.types'
import KeyValueMultiLineListItem from './KeyValueMultiLineListItem'
import type { KeyValueListItemActionProps } from './common'

import type { KeyValueListItemContainerProps } from './KeyValueListItem.common.styled'

export type {
  KeyValueListItemContainerProps,
  KeyValueListItemActionProps,
  KeyValueListItemProps,
}

export {
  KeyValueListItem,
  KeyValueMultiLineListItem,
}
