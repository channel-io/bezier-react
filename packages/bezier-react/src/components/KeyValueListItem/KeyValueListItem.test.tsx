import React from 'react'

import {
  AppleIcon,
  BadgeIcon,
} from '@channel.io/bezier-icons'
import userEvent from '@testing-library/user-event'

import { render } from '~/src/utils/testUtils'

import KeyValueListItem from './KeyValueListItem'
import { TEST_ID_MAP } from './KeyValueListItem.const'
import { type KeyValueListItemProps } from './KeyValueListItem.types'
import KeyValueMultiLineListItem from './KeyValueMultiLineListItem'
import type { KeyValueListItemActionProps } from './common'

const DEFAULT_PROPS: KeyValueListItemProps = {
  keyIcon: AppleIcon,
  keyContent: 'Key',
  children: 'Value',
}

const renderComponent = (optionProps?: Partial<KeyValueListItemProps>) => render(
  <KeyValueListItem {...DEFAULT_PROPS} {...optionProps} />,
)

const renderMultilineComponent = (optionProps?: Partial<KeyValueListItemProps>) => render(
  <KeyValueMultiLineListItem {...DEFAULT_PROPS} {...optionProps} />,
)

describe('KeyValueListItem', () => {
  it('Snapshot >', () => {
    const { getByTestId } = renderComponent()
    const rendered = getByTestId(TEST_ID_MAP.ROOT)
    expect(rendered).toMatchSnapshot()
  })

  describe('Props', () => {
    describe('keyIcon', () => {
      it('keyIcon이 Bezier의 아이콘 한 종류면, icon으로 렌더링된다.', () => {
        const keyIcon = BadgeIcon
        const { getByTestId } = renderComponent({ keyIcon })
        const rendered = getByTestId(TEST_ID_MAP.KEY_ITEM)
        const keyItemIcon = rendered?.firstChild
        expect(keyItemIcon?.nodeName).toEqual('svg')
      })

      it('keyIcon이 Bezier의 아이콘이 아니면, textContent로 렌더링된다.', () => {
        const keyIcon = 'NoIcon'
        const { getByTestId } = renderComponent({ keyIcon })
        const rendered = getByTestId(TEST_ID_MAP.KEY_ITEM)
        const keyItemIcon = rendered?.firstChild
        expect(keyItemIcon?.nodeName).toEqual('#text')
        expect(keyItemIcon?.textContent).toEqual(keyIcon)
      })
    })

    describe('keyContent', () => {
      it('keyContent가 string이면, Key의 text string으로서 렌더링된다.', () => {
        const { getByTestId } = renderComponent()
        const rendered = getByTestId(TEST_ID_MAP.KEY_ITEM)
        const keyItemText = rendered?.lastChild
        expect(keyItemText?.nodeName).toEqual('DIV')
        expect(keyItemText?.textContent).toEqual(DEFAULT_PROPS.keyContent)
      })

      it('keyContent가 React.Node면, Key의 text가 node로서 렌더링된다.', () => {
        const keyContent = (
          <button type="button">
            Button
          </button>
        )
        const { getByTestId } = renderComponent({ keyContent })
        const rendered = getByTestId(TEST_ID_MAP.KEY_ITEM)
        const keyItemText = rendered?.lastChild
        expect(keyItemText?.nodeName).toEqual('BUTTON')
        expect(keyItemText?.textContent).toEqual('Button')
      })
    })

    describe('children', () => {
      it('children가 string이면, Value의 text string으로서 렌더링된다.', () => {
        const { getByTestId } = renderComponent()
        const rendered = getByTestId(TEST_ID_MAP.VALUE_ITEM)
        expect(rendered?.nodeName).toEqual('DIV')
        expect(rendered?.textContent).toEqual(DEFAULT_PROPS.children)
      })

      it('children가 React.Node면, Value의 text가 node로서 렌더링된다.', () => {
        const children = (
          <button type="button">
            Button
          </button>
        )
        const { getByTestId } = renderComponent({ children })
        const rendered = getByTestId(TEST_ID_MAP.VALUE_ITEM)
        const valueChildren = rendered?.firstChild
        expect(valueChildren?.nodeName).toEqual('BUTTON')
        expect(valueChildren?.textContent).toEqual('Button')
      })
    })

    describe('actions', () => {
      it('actions의 show가 true면, ActionIconWrapper display가 flex처리 된다.', () => {
        const actions: KeyValueListItemActionProps = { icon: BadgeIcon, show: true }
        const { getByTestId } = renderComponent({ actions })
        const rendered = getByTestId(TEST_ID_MAP.ACTIONS_ITEM)
        const actionItemIconWrapper = rendered?.firstChild
        expect(actionItemIconWrapper).toHaveStyle('display: flex;')
      })

      it('actions의 show가 false면, ActionIconWrapper display가 none처리 된다.', () => {
        const actions: KeyValueListItemActionProps = { icon: BadgeIcon, show: false }
        const { getByTestId } = renderComponent({ actions })
        const rendered = getByTestId(TEST_ID_MAP.ACTIONS_ITEM)
        const actionItemIconWrapper = rendered?.firstChild
        expect(actionItemIconWrapper).toHaveStyle('display: none;')
      })

      it('actions가 없으면, ItemAction은 렌더링되지 않는다.', async () => {
        const { queryByTestId } = renderComponent({ })
        const rendered = await queryByTestId(TEST_ID_MAP.ACTIONS_ITEM)
        const actionItems = rendered
        expect(actionItems).not.toBeInTheDocument()
      })

      it('actions가 Object면, 아이콘이 1개만 렌더링된다.', () => {
        const actions: KeyValueListItemActionProps = { icon: BadgeIcon }
        const { getByTestId } = renderComponent({ actions })
        const rendered = getByTestId(TEST_ID_MAP.ACTIONS_ITEM)
        const actionItemsCount = rendered?.childNodes.length
        expect(actionItemsCount).toEqual(1)
      })

      it('actions가 Array면, 아이콘이 여러개 렌더링된다.', () => {
        const actions: KeyValueListItemActionProps[] = (
          Array.from(Array(2)).map(() => ({ icon: BadgeIcon }))
        )
        const { getByTestId } = renderComponent({ actions })
        const rendered = getByTestId(TEST_ID_MAP.ACTIONS_ITEM)
        const actionItemsCount = rendered?.childNodes.length
        expect(actionItemsCount).toEqual(2)
      })

      it('actions의 icon이 렌더링된다.', () => {
        const actions: KeyValueListItemActionProps = { icon: BadgeIcon }
        const { getByTestId } = renderComponent({ actions })
        const rendered = getByTestId(TEST_ID_MAP.ACTIONS_ITEM)
        const actionItemIconWrapper = rendered?.firstChild
        const actionItemIcon = actionItemIconWrapper?.firstChild
        expect(actionItemIcon?.nodeName).toEqual('svg')
      })

      it('actions의 onClick 있으면, icon click event에서 호출된다.', async () => {
        const user = userEvent.setup()

        const actions: KeyValueListItemActionProps = { icon: BadgeIcon, onClick: jest.fn() }
        const { getByTestId } = renderComponent({ actions })
        const rendered = getByTestId(TEST_ID_MAP.ACTIONS_ITEM)
        const actionItemIconWrapper = rendered?.firstChild
        const actionItemIcon = actionItemIconWrapper?.firstChild

        await user.click(actionItemIcon as HTMLElement)
        expect(actions.onClick).toBeCalledTimes(1)
        await user.click(actionItemIcon as HTMLElement)
        expect(actions.onClick).toBeCalledTimes(2)
      })

      it('actions의 tooltip이 있으면, tooltip이 렌더링된다.', async () => {
        const user = userEvent.setup()

        const actions: KeyValueListItemActionProps = { icon: BadgeIcon, tooltip: 'tooltip' }
        const { getByRole, getByTestId } = renderComponent({ actions })
        const rendered = getByTestId(TEST_ID_MAP.ACTIONS_ITEM)
        const actionItemIconWrapper = rendered?.firstChild
        const actionItemIcon = actionItemIconWrapper?.firstChild

        await user.hover(actionItemIcon as HTMLElement)
        expect(getByRole('tooltip')).toBeInTheDocument()
      })
    })

    describe('onClickKey', () => {
      it('onClickKey가 없으면, Key 영역의 hover/cursor 스타일이 없다.', async () => {
        const user = userEvent.setup()

        const { getByTestId } = renderComponent()
        const rendered = getByTestId(TEST_ID_MAP.ROOT)
        const keyItemContainer = rendered?.firstChild

        await user.hover(keyItemContainer as Element)
        expect(keyItemContainer).toHaveStyle('display: flex;')
        expect(keyItemContainer).toHaveStyle('align-items: flex-start;')
        expect(keyItemContainer).toHaveStyle('padding: 4px 6px;')
        expect(keyItemContainer).not.toHaveStyle('cursor: pointer;')
        // NOTE(@sol): Hovered DOM style test is not working
        // expect(keyItemContainer).not.toHaveStyle(`background-color: rgba(0, 0, 0, 0.051);`)
      })

      it('onClickKey가 있으면, Key 영역이 hover/cursor 스타일이 존재한다.', async () => {
        const user = userEvent.setup()

        const props: Partial<KeyValueListItemProps> = {
          onClickKey: jest.fn(),
        }
        const { getByTestId } = renderComponent(props)
        const rendered = getByTestId(TEST_ID_MAP.ROOT)
        const keyItemContainer = rendered?.firstChild

        await user.hover(keyItemContainer as Element)
        expect(keyItemContainer).toHaveStyle('display: flex;')
        expect(keyItemContainer).toHaveStyle('align-items: flex-start;')
        expect(keyItemContainer).toHaveStyle('padding: 4px 6px;')
        expect(keyItemContainer).toHaveStyle('cursor: pointer;')
        // NOTE(@sol): Hovered DOM style test is not working
        // expect(keyItemContainer).toHaveStyle(`background-color: rgba(0, 0, 0, 0.051);`)
      })

      it('onClickKey가 있으면, Key 영역이 click event에서 호출된다.', async () => {
        const user = userEvent.setup()

        const props: Partial<KeyValueListItemProps> = {
          onClickKey: jest.fn(),
        }
        const { getByTestId } = renderComponent(props)
        const rendered = getByTestId(TEST_ID_MAP.ROOT)
        const keyItemContainer = rendered?.firstChild

        await user.click(keyItemContainer as Element)
        expect(props.onClickKey).toBeCalledTimes(1)
        await user.click(keyItemContainer as Element)
        expect(props.onClickKey).toBeCalledTimes(2)
      })
    })

    describe('onClickValue', () => {
      it('onClickValue가 없으면, Value 영역의 hover/cursor 스타일이 없다.', async () => {
        const user = userEvent.setup()

        const { getByTestId } = renderComponent()
        const rendered = getByTestId(TEST_ID_MAP.ROOT)
        const valueItemContainer = rendered?.lastChild

        await user.hover(valueItemContainer as Element)
        expect(valueItemContainer).toHaveStyle('display: flex;')
        expect(valueItemContainer).toHaveStyle('align-items: flex-start;')
        expect(valueItemContainer).toHaveStyle('padding: 4px 6px;')
        expect(valueItemContainer).not.toHaveStyle('cursor: pointer;')
        // NOTE(@sol): Hovered DOM style test is not working
        // expect(valueItemContainer).not.toHaveStyle(`background-color: rgba(0, 0, 0, 0.051);`)
      })

      it('onClickValue가 있으면, Value 영역이 hover/cursor 스타일이 존재한다.', async () => {
        const user = userEvent.setup()

        const props: Partial<KeyValueListItemProps> = {
          onClickValue: jest.fn(),
        }
        const { getByTestId } = renderComponent(props)
        const rendered = getByTestId(TEST_ID_MAP.ROOT)
        const valueItemContainer = rendered?.lastChild

        await user.hover(valueItemContainer as Element)
        expect(valueItemContainer).toHaveStyle('display: flex;')
        expect(valueItemContainer).toHaveStyle('align-items: flex-start;')
        expect(valueItemContainer).toHaveStyle('padding: 4px 6px;')
        expect(valueItemContainer).toHaveStyle('cursor: pointer;')
        // NOTE(@sol): Hovered DOM style test is not working
        // expect(valueItemContainer).toHaveStyle('background-color: rgba(0, 0, 0, 0.051);')
      })

      it('onClickValue가 있으면, Value 영역이 click event에서 호출된다.', async () => {
        const user = userEvent.setup()

        const props: Partial<KeyValueListItemProps> = {
          onClickValue: jest.fn(),
        }
        const { getByTestId } = renderComponent(props)
        const rendered = getByTestId(TEST_ID_MAP.ROOT)
        const valueItemContainer = rendered?.lastChild

        await user.click(valueItemContainer as Element)
        expect(props.onClickValue).toBeCalledTimes(1)
        await user.click(valueItemContainer as Element)
        expect(props.onClickValue).toBeCalledTimes(2)
      })
    })
  })

  describe('Styles', () => {
    describe('KeyValueListItem', () => {
      it('Root', () => {
        const { getByTestId } = renderComponent()
        const rendered = getByTestId(TEST_ID_MAP.ROOT)
        expect(rendered).toHaveStyle('display: flex;')
        expect(rendered).toHaveStyle('align-items: flex-start;')
        expect(rendered).toHaveStyle('border-radius: 6px;')
        expect(rendered).toHaveStyle('transition-delay: 0ms;')
        expect(rendered).toHaveStyle('transition-timing-function: cubic-bezier(.3,0,0,1);')
        expect(rendered).toHaveStyle('transition-duration: 150ms;')
        expect(rendered).toHaveStyle('transition-property: background-color,color;')

        const rootKeyContainer = rendered.firstChild
        expect(rootKeyContainer).toHaveStyle('display: flex;')
        expect(rootKeyContainer).toHaveStyle('flex: 1;')

        const rootValueContainer = rendered.lastChild
        expect(rootValueContainer).toHaveStyle('display: flex;')
        expect(rootValueContainer).toHaveStyle('flex: 2;')
        expect(rootValueContainer).toHaveStyle('justify-content: space-between;')
        expect(rootValueContainer).toHaveStyle('overflow: hidden;')
        expect(rootValueContainer).toHaveStyle('text-overflow: ellipsis;')
        expect(rootValueContainer).toHaveStyle('white-space: nowrap;')
      })

      it('Key, Value가 flex로서 1:2의 width 비율을 유지한다.', () => {
        const { getByTestId } = renderComponent()
        const rendered = getByTestId(TEST_ID_MAP.ROOT)

        const rootKeyContainer = rendered.firstChild
        expect(rootKeyContainer).toHaveStyle('display: flex;')
        expect(rootKeyContainer).toHaveStyle('flex: 1;')

        const rootValueContainer = rendered.lastChild
        expect(rootValueContainer).toHaveStyle('display: flex;')
        expect(rootValueContainer).toHaveStyle('flex: 2;')
        expect(rootValueContainer).toHaveStyle('justify-content: space-between;')
      })

      it('KeyValueListItem > KeyItem', () => {
        const { getByTestId } = renderComponent()
        const rendered = getByTestId(TEST_ID_MAP.KEY_ITEM)
        expect(rendered).toHaveStyle('display: flex;')
        expect(rendered).toHaveStyle('align-items: center;')

        const keyItemText = rendered?.lastChild
        expect(keyItemText).toHaveStyle('color: rgba(0, 0, 0, 0.4);')
        expect(keyItemText).toHaveStyle('overflow: hidden;')
        expect(keyItemText).toHaveStyle('text-overflow: ellipsis;')
        expect(keyItemText).toHaveStyle('white-space: nowrap;')
      })

      it('KeyValueListItem > ValueItem', () => {
        const { getByTestId } = renderComponent()
        const rendered = getByTestId(TEST_ID_MAP.VALUE_ITEM)
        expect(rendered).toHaveStyle('font-size: 1.4rem;')
      })

      it('KeyValueListItem > ItemActions', () => {
        const actions: KeyValueListItemActionProps = { icon: BadgeIcon }
        const { getByTestId } = renderComponent({ actions })
        const rendered = getByTestId(TEST_ID_MAP.ACTIONS_ITEM)
        expect(rendered).toHaveStyle('display: flex;')
        expect(rendered).toHaveStyle('align-items: center;')
      })
    })

    describe('KeyValueMultiLineListItem', () => {
      it('Root', () => {
        const { getByTestId } = renderMultilineComponent()
        const rendered = getByTestId(TEST_ID_MAP.MULTILINE_ROOT)
        expect(rendered).toHaveStyle('display: flex;')
        expect(rendered).toHaveStyle('flex-direction: column;')
        expect(rendered).toHaveStyle('justify-content: center;')
        expect(rendered).toHaveStyle('border-radius: 6px;')
        expect(rendered).toHaveStyle('transition-delay: 0ms;')
        expect(rendered).toHaveStyle('transition-timing-function: cubic-bezier(.3,0,0,1);')
        expect(rendered).toHaveStyle('transition-duration: 150ms;')
        expect(rendered).toHaveStyle('transition-property: background-color,color;')

        const rootKeyContainer = rendered.firstChild
        expect(rootKeyContainer).toHaveStyle('display: flex;')
        expect(rootKeyContainer).toHaveStyle('flex: 1;')
        expect(rootKeyContainer).toHaveStyle('justify-content: space-between;')
        expect(rootKeyContainer).toHaveStyle('height: 28px;')

        const rootValueContainer = rendered.lastChild
        expect(rootValueContainer).toHaveStyle('display: flex;')
        expect(rootValueContainer).toHaveStyle('flex: 1;')
        expect(rootValueContainer).toHaveStyle('padding: 4px 6px;')
        expect(rootValueContainer).toHaveStyle('min-height: 28px;')
      })
    })
  })
})
