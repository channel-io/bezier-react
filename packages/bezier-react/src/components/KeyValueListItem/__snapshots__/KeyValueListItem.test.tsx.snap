// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`KeyValueListItem Snapshot > 1`] = `
.c5 {
  -webkit-flex: 0 0 auto;
  -ms-flex: 0 0 auto;
  flex: 0 0 auto;
  margin: 0px 0px 0px 0px;
  color: #00000066;
  -webkit-transition-delay: 0ms;
  transition-delay: 0ms;
  -webkit-transition-timing-function: cubic-bezier(.3,0,0,1);
  transition-timing-function: cubic-bezier(.3,0,0,1);
  -webkit-transition-duration: 150ms;
  transition-duration: 150ms;
  -webkit-transition-property: color;
  transition-property: color;
}

.c6 {
  font-size: 1.2rem;
  line-height: 1.6rem;
  display: block;
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
  margin: 0px 0px 0px 0px;
  font-style: normal;
  font-weight: bold;
  color: #00000066;
  -webkit-transition-delay: 0ms;
  transition-delay: 0ms;
  -webkit-transition-timing-function: cubic-bezier(.3,0,0,1);
  transition-timing-function: cubic-bezier(.3,0,0,1);
  -webkit-transition-duration: 150ms;
  transition-duration: 150ms;
  -webkit-transition-property: color;
  transition-property: color;
}

.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
}

.c7 {
  margin-left: 8px;
}

.c9 {
  color: #000000D9;
  font-size: 1.4rem;
  line-height: 1.8rem;
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
}

.c0 {
  overflow: hidden;
  border-radius: 6px;
  -webkit-transition-delay: 0ms;
  transition-delay: 0ms;
  -webkit-transition-timing-function: cubic-bezier(.3,0,0,1);
  transition-timing-function: cubic-bezier(.3,0,0,1);
  -webkit-transition-duration: 150ms;
  transition-duration: 150ms;
  -webkit-transition-property: background-color,color;
  transition-property: background-color,color;
}

.c2 {
  box-sizing: border-box;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
  padding: 4px 6px;
  overflow: hidden;
  border-radius: 6px;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
}

.c3 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
}

.c8 {
  -webkit-flex: 2;
  -ms-flex: 2;
  flex: 2;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  min-height: 28px;
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
}

<div
  class="c0 c1"
  data-testid="bezier-react-key-value-list-item"
>
  <div
    class="c2 c3"
  >
    <div
      class="c4"
      data-testid="bezier-react-key-value-list-item-key-item"
    >
      <svg
        class="c5"
        color="txt-black-dark"
        data-testid="bezier-react-icon"
        fill="none"
        foundation="[object Object]"
        height="20"
        marginbottom="0"
        marginleft="0"
        marginright="0"
        margintop="0"
        viewBox="0 0 24 24"
        width="20"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          clip-rule="evenodd"
          d="M15.832 1s.272 1.628-1.033 3.197c-1.394 1.675-2.979 1.4-2.979 1.4s-.297-1.317.872-2.857C14.006 1.01 15.832 1 15.832 1m3.637 6.81s-2.163 1.105-2.163 3.789c0 3.027 2.694 4.07 2.694 4.07s-1.883 5.302-4.427 5.302c-.577 0-1.09-.192-1.611-.387-.534-.2-1.075-.401-1.698-.401-.695 0-1.388.25-2.005.475-.497.18-.945.342-1.306.342-2.322 0-5.257-5.03-5.257-9.071 0-3.977 2.484-6.063 4.814-6.063.88 0 1.646.295 2.288.542.462.178.86.33 1.19.33.264 0 .618-.142 1.047-.315.669-.27 1.522-.613 2.517-.613 2.81 0 3.917 2 3.917 2"
          fill="currentColor"
          fill-rule="evenodd"
        />
      </svg>
      <div
        class="c6 c7"
        color="txt-black-dark"
        data-testid="bezier-react-text"
      >
        Key
      </div>
    </div>
  </div>
  <div
    class="c2 c8"
  >
    <div
      class="c9"
      data-testid="bezier-react-key-value-list-item-value-item"
    >
      Value
    </div>
  </div>
</div>
`;
